compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala

```diff
@@ -528,4 +546,27 @@ object GenericSignatures {
     val rte = recur(mtd)
     (tparams.toList, vparams.toList, rte)
   end collectMethodParams
+
+  /** Collect type parameters that are actually used in the given types. */
+  private def collectUsedTypeParams(types: List[Type], initialSymbol: Symbol)(using Context): (Set[Name], Set[Symbol]) =
+    val usedMethodTypeParamNames = collection.mutable.Set.empty[Name]
+    val usedClassTypeParams = collection.mutable.Set.empty[Symbol]
+    val collector = new TypeTraverser:
+      def traverse(tp: Type) = tp.dealias match
+        case ref @ TypeParamRef(_: PolyType, _) =>
+          usedMethodTypeParamNames += ref.paramName
+        case TypeRef(pre, _) =>
+          val sym = tp.typeSymbol
+          if isTypeParameterInMethSig(sym, initialSymbol) then
+            usedMethodTypeParamNames += sym.name
+          else if sym.isTypeParam && sym.isContainedIn(initialSymbol.topLevelClass) then
+            usedClassTypeParams += sym
+          else
+            traverse(pre)
```

Comment:
this should probably not be in the `else` branch?

Reply:
Right, we don't need to traverse the `pre`